lib/rpmsg/rpmsg_virtio.c

```diff
@@ -262,14 +263,128 @@ static int _rpmsg_virtio_get_buffer_size(struct rpmsg_virtio_device *rvdev)
 	return length;
 }
 
+static void rpmsg_virtio_hold_rx_buffer(struct rpmsg_device *rdev, void *rxbuf)
+{
+	struct rpmsg_hdr *rp_hdr;
+
+	(void)rdev;
+
+	rp_hdr = RPMSG_LOCATE_HDR(rxbuf);
+
+	/* Set held status to keep buffer */
+	rp_hdr->reserved = RPMSG_BUF_HELD;
+}
+
+static void rpmsg_virtio_release_rx_buffer(struct rpmsg_device *rdev, void *rxbuf)
+{
+	struct rpmsg_virtio_device *rvdev;
+	struct rpmsg_hdr *rp_hdr;
+	uint16_t idx;
+	uint32_t len;
+
+	rvdev = metal_container_of(rdev, struct rpmsg_virtio_device, rdev);
+	rp_hdr = RPMSG_LOCATE_HDR(rxbuf);
+	/* The reserved field contains buffer index */
+	idx = rp_hdr->reserved;
+
+	metal_mutex_acquire(&rdev->lock);
+	/* Return buffer on virtqueue. */
+	len = virtqueue_get_buffer_length(rvdev->rvq, idx);
```

Comment:
if it is the virtio master, you cannot keep the idx for returning the buffer.
It will be ok for slave. On the other hand, I think it ok to assuming that the buffer size are always the same for master RX.
I am ok with this change. Please add comments to this line to indicate you can only use the idx to locate the old buffer if it is  virtio slave,  on the master side, the buffer sizes on the RX are the same.

Reply:
> if it is the virtio master, you cannot keep the idx for returning the buffer.

It's depend on where the buffer come from:
1.The idx is invalid if the buffer come from share memory buffer pool
2.The idx is valid if the buffer come from tx used queue

> It will be ok for slave. On the other hand, I think it ok to assuming that the buffer size are always the same for master RX.

Yes, because these two points:
1.idx isn't always valid for master side
2.the buffer size could be assumed same in master side
The previous patch remove the buffer length argument.

> I am ok with this change. Please add comments to this line to indicate you can only use the idx to locate the old buffer if it is virtio slave, on the master side, the buffer sizes on the RX are the same.

Ok.

Reply:
> The previous patch remove the buffer length argument.

I think the len argument should be there, just if it is slave, as it is not the one to allocate the buffer, it cannot control the length of the buffer, it can get it from the index. For master, it depends on allocator, and i think usually, it is a fixed size.